File: leetcode/21-merge-two-sorted-lists/main.go
package main

import "sort"

type ListNode struct {
	Val  int
	Next *ListNode
}

/*
	0th approach: naive way
	- iterate the list and put all the values into an array
	- sort the array
	- make that array into a linked list
	Time	O(NlogN) N=len(l1)+len(l2)
	Space	O(N)
	4ms beats 100%
*/
func mergeTwoLists0(l1 *ListNode, l2 *ListNode) *ListNode {
	a := []int{}
	cur := l1
	for cur != nil {
		a = append(a, cur.Val)
		cur = cur.Next
	}
	cur = l2
	for cur != nil {
		a = append(a, cur.Val)
		cur = cur.Next
	}
	sort.Ints(a)
	dumpHead := &ListNode{0, nil}
	cur = dumpHead
	for i := 0; i < len(a); i++ {
		cur.Next = &ListNode{a[i], nil}
		cur = cur.Next
	}
	return dumpHead.Next
}

/*
	1st approach
	- idea like merge sort
	Time	O(N)
	Space	O(N)
	4ms beats 100%
*/
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	curA := l1
	curB := l2
	dump := &ListNode{0, nil}
	cur := dump
	for curA != nil && curB != nil {
		if curA.Val < curB.Val {
			cur.Next = curA
			curA = curA.Next
		} else {
			cur.Next = curB
			curB = curB.Next
		}
		cur = cur.Next
	}
	for curA != nil {
		cur.Next = curA
		curA = curA.Next
		cur = cur.Next
	}
	for curB != nil {
		cur.Next = curB
		curB = curB.Next
		cur = cur.Next
	}
	// append intermediate result to result
	return dump.Next
}

func main() {

}

File: leetcode/21-merge-two-sorted-lists/main.py
"""
    Merge two sorted linked lists and return it as a new list. 
    The new list should be made by splicing together the nodes of the first two lists.

    Time    O(A+B)
    Space   O(A+B)
    28 ms, faster than 64.53%
"""


class ListNode(object):
    def __init__(self, val):
        self.val = val
        self.next = None


class Solution(object):
    def merge2Lists(self, l1, l2):
        """
            Time
            24ms beats 100%
        """
        dumphead = ListNode()
        cur = dumphead
        cur1 = l1
        cur2 = l2
        while cur1 and cur2:
            if cur1.val < cur2.val:
                cur.next = ListNode(cur1.val)
                cur1 = cur1.next
            else:
                cur.next = ListNode(cur2.val)
                cur2 = cur2.next
            cur = cur.next
        if cur1:
            cur.next = cur1
        elif cur2:
            cur.next = cur2
        return dumphead.next


# helpers for test
def arr2list(arr):
    dump = ListNode(0)
    dumpCur = dump
    for i in range(len(arr)):
        dumpCur.next = ListNode(arr[i])
        dumpCur = dumpCur.next
    return dump.next


def printList(l):
    c = l
    acc = ""
    while c != None:
        acc += str(c.val) + ','
        c = c.next
    print(acc)


a = arr2list([1, 3, 5])
b = arr2list([2, 4, 6])
s = Solution().merge2Lists(a, b)
printList(s)

a = arr2list([1, 3, 5])
b = arr2list([2, 4, 5, 6])
s = Solution().merge2Lists(a, b)
printList(s)
